src/io/ioexternalbus.cpp

```diff
@@ -28,10 +50,217 @@ QString IOExternalBus::description() const {
+
+                    m_ui->connectButton->setText("Disconnect");
+                    m_ui->status->setText("Connected");
+                    m_ui->server->setText(desc.object().value(QString("name")).toString());
+
+                    emit regMapChanged();
+                    emit sizeChanged();
+
+                    // delete[] buff;
+                } else {
+                    QMessageBox::information(nullptr, tr("Ripes VBus"), QString("json: ") + error.errorString());
+                    m_ui->connectButton->setText("Connect");
+                    m_ui->status->setText("Disconnected");
+                    m_ui->server->setText("-");
+
+                    send_cmd(VB_QUIT);
+                    tcpSocket->abort();
+                }
+            }
+        } else {
+            disconnectOnError();
+        }
+    } else {  // disconnect
+        m_ui->connectButton->setText("Connect");
+        m_ui->status->setText("Disconnected");
+        m_ui->server->setText("-");
+
+        send_cmd(VB_QUIT);
+        tcpSocket->abort();
+
+        m_regDescs.clear();
+
+        emit regMapChanged();
+        emit sizeChanged();
+    }
+}
+
+int32_t IOExternalBus::send_cmd(const uint32_t cmd, const uint32_t payload_size, const QByteArray* payload) {
+    int32_t ret = -1;
+    cmd_header_t cmd_header;
+
+    cmd_header.msg_type = htonl(cmd);
+    cmd_header.payload_size = htonl(payload_size);
+
+    QByteArray dp = QByteArray(static_cast<const char*>((void*)&cmd_header), sizeof(cmd_header_t));
+
+    if (payload_size) {
+        dp.append(*payload);
+    }
+
+    if ((ret = tcpSocket->write(dp, dp.size())) < 0) {
+        disconnectOnError();
+        return -1;
+    }
+
+    return ret;
 }
-void IOExternalBus::ioWrite(AInt offset, VInt value, unsigned size) {}
 
-void IOExternalBus::parameterChanged(unsigned paramId) {}
+int32_t IOExternalBus::recv_cmd(cmd_header_t& cmd_header) {
+    QByteArray dp = QByteArray(static_cast<char*>((void*)&cmd_header), sizeof(cmd_header_t));
+    int ret = tcpSocket->read(dp, sizeof(cmd_header_t));
+    if (ret < 0) {
+        disconnectOnError();
+        return -1;
+    }
+
+    cmd_header_t* hr = (cmd_header_t*)dp.data();
+
+    cmd_header.msg_type = ntohl(hr->msg_type);
+    cmd_header.payload_size = ntohl(hr->payload_size);
+
+    return ret;
+}
+
+int32_t IOExternalBus::recv_payload(QByteArray& buff, const uint32_t payload_size) {
+    int ret = tcpSocket->read(buff, payload_size);
+    if (ret < 0) {
+        disconnectOnError();
+        return -1;
+    }
+    return ret;
+}
+
+void IOExternalBus::disconnectOnError(void) {
+    skt_use = 0;
+    QMessageBox::information(nullptr, tr("Ripes VBus"), tcpSocket->getLastErrorStr());
+    tcpSocket->close();
+
+    m_ui->connectButton->setText("Connect");
+    m_ui->status->setText("Disconnected");
+    m_ui->server->setText("-");
+    tcpSocket->abort();
+
+    m_regDescs.clear();
+
+    emit regMapChanged();
+    emit sizeChanged();
```

Comment:
Since this appears everywhere where we update `m_ui->...` things, could be factored into the `updateConnectionStatus` function proposed above.

Reply:
Done.